bridge/src/xcm_asset_transactor.rs

```diff
@@ -0,0 +1,81 @@
+use core::marker::PhantomData;
+
+use codec::Encode;
+use xcm::latest::{Junction, MultiAsset, MultiLocation, XcmContext};
+use xcm::prelude::*;
+use xcm_executor::{Assets, traits::TransactAsset};
+use hex_literal::hex;
+use sygma_traits::{AssetTypeIdentifier, TransactorForwarder};
+
+pub struct XCMAssetTransactor<CurrencyTransactor, FungiblesTransactor, AssetTypeChecker, Forwarder>(PhantomData<(CurrencyTransactor, FungiblesTransactor, AssetTypeChecker, Forwarder)>);
+
+impl<CurrencyTransactor: TransactAsset, FungiblesTransactor: TransactAsset, AssetTypeChecker: AssetTypeIdentifier, Forwarder: TransactorForwarder> TransactAsset for XCMAssetTransactor<CurrencyTransactor, FungiblesTransactor, AssetTypeChecker, Forwarder> {
+    // deposit_asset implements the TransactAsset deposit_asset method and contains the logic to classify
+    // the asset recipient location:
+    // 1. recipient is on the local parachain
+    // 2. recipient is on non-substrate chain(evm, cosmos, etc.)
+    // 3. recipient is on the remote parachain
+    fn deposit_asset(what: &MultiAsset, who: &MultiLocation, context: &XcmContext) -> XcmResult {
+        match (who.parents, who.first_interior()) {
+            // 1. recipient is on the local parachain
+            (0, Some(Parachain(_))) => {
+                // check if the asset is native, and call the corresponding deposit_asset()
+                if AssetTypeChecker::is_native_asset(what) {
+                    CurrencyTransactor::deposit_asset(what, who, context)?;
+                } else {
+                    FungiblesTransactor::deposit_asset(what, who, context)?
+                }
+            }
+            // recipient is on the remote chain
+            (1, Some(Parachain(_))) => {
```

Comment:
And here I suppose the full dest was set to `(1, x5[Parachain(your parachain id), ...]` on source parachain, then here the pattern should be like `(0, X4[...])` without your parachain ID since the message currently already executing on your parachain

Reply:
that makes a lot sense